### PR TITLE
Convert MKL symbols from global to local

### PR DESCRIPTION
**Pull request by CuiYifeng**
PyTorch is statically linked to MKL but MKL symbols are visible to global, which may cause symbol conflicts.
Such error has been observed when a different version of MKL is dynamically linked to the other components: `libtorch_cpu.so` was invoked incorrectly when MKL descriptor was freed.

### CODE REVIEW

File: cmake/public/mkl.cmake
Comment (EikanWang) [inline cmake/public/mkl.cmake:26]:
```suggestion
  if(USE_STATIC_MKL)
```

**Review by EikanWang — APPROVED** [review]

**Comment by cyyever** [comment]
I had similar trails before, but some tests kept failing.

**Comment by EikanWang** [comment]
> I had similar trails before, but some tests kept failing.

Thanks for the information. Are the failed tests part of the CI? We can try to triage the CI failed test cases first.

**Comment by cyyever** [comment]
@CuiYifeng Some CI tests would fail by a large chance for missing LAPACK symbols.

**Comment by CuiYifeng** [comment]
> @CuiYifeng Some CI tests would fail by a large chance for missing LAPACK symbols. And I'm not sure what would happen (random crashes?) when your system uses two different versions of MKL, one is statically linked to PyTorch after this PR. The idea solution is using one and only one version of MKL consistently throughout your system.
> 
> See also the comment starting in line 1423 of caffe2/CMakeLists.txt. The interaction becomes even trickier with the introduction of MAGMA. That is why I gave up after several days of fighting with MKL, LAPACK and MAGMA, dealing with various linking issues.

@cyyever Thanks for your information. Yes, there will be random crashes when we use two different versions of MKL. According to the comments you mentioned, the fundamental motivation for retaining extra MKL symbols in torch_cpu is to save 8M binary size for torch_cuda, right?
So far, I have only seen failures in Windows cases. This PR should not have any impact on Windows. Let's take a look after CI is over.

**Review by cyyever — APPROVED** [review]

**Comment by CuiYifeng** [comment]
Why is it displaying `All checks have passed` but CI still shows "6 New Failures" in comments?

**Comment by CuiYifeng** [comment]
> > Why is it displaying `All checks have passed` but CI still shows "6 New Failures" in comments?
> 
> You can merge now, the failures were collected yesterday and sometimes they are not updated for some reason.

Thanks for the information.
@EikanWang Please help merge.

**Comment by EikanWang** [comment]
@malfet , we cannot merge this PR as it requires review from core maintainers. May I know if you can help review this PR?

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**
